### PR TITLE
Break the initialization cycle in NIO/cgroups code

### PR DESCRIPTION
**Pull request by jerboaa**
When building with an unpatched JDK 21, there is an initialization cycle in the native image code. The details are explained in this issue:
https://github.com/graalvm/mandrel/issues/556

This isn't easily reproducible with a LabsJDK 21 based build. However, I think it would be good to not be dependent on the patch in the base JDK. We have been using this patch in Mandrel as we are using unpatched OpenJDK 21 as a base.

**Details of the patch:**

First, we use a separate accessor for page-alignedness as it doesn't need the more sophisticated initialization of the directMemory field.

Next, ensure PhysicalMemory initialization is serialized and when it is, set directMemory to a static value so that the container code can finish initialization without introducing a cyle. The final directMemory value based on the heap size is then published to JDK code by setting the VM init level to 1. Therefore, application code would use the non-static value as the upper bound.

Closes: https://github.com/graalvm/mandrel/issues/556

### CODE REVIEW

**Comment by jerboaa** [comment]
This is an alternative fix to:
https://github.com/graalvm/labs-openjdk-21/commit/db1723c5fbeae563339e322d9794971803d183b4

**Comment by jerboaa** [comment]
@adinn @roberttoyonaga Could you perhaps review this, please?

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/PhysicalMemory.java [hunk 1]
```diff
@@ -80,13 +94,26 @@ public static UnsignedWord size() {
...
+                    /*
+                     * Recursive initializations need to wait for the one initializing thread to
+                     * finish so as to get correct reads of the cachedSize value.
+                     */
+                    try {
+                        boolean expired = !CACHED_SIZE_AVAIL_LATCH.await(1L, TimeUnit.SECONDS);
+                        if (expired) {
+                            throw new InternalError("Expired latch!");
+                        }
+                        VMError.guarantee(cachedSize != UNSET_SENTINEL, "Expected chached size to be set");
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/PhysicalMemory.java:110]: Thanks!

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_internal_misc_VM.java [hunk 2]
```diff
@@ -107,17 +118,82 @@ private static void initialize() {
...
+        VMError.guarantee(newDirectMemory > 0, "New direct memory should be initialized");
 
-        /*
-         * The initialization is not synchronized, so multiple threads can race. Usually this will
-         * lead to the same value, unless the runtime options are modified concurrently - which is
-         * possible but not a case we care about.
-         */
+        Unsafe.getUnsafe().storeFence();
         directMemory = newDirectMemory;
+        if (PhysicalMemory.isInitialized()) {
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_internal_misc_VM.java]: Yes, good point. The update patch should no longer have that problem.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_internal_misc_VM.java [hunk 3]
```diff
@@ -67,36 +70,44 @@ public static ClassLoader latestUserDefinedLoader0() {
...
 
-    static boolean getPageAlignDirectMemory() {
-        if (!initialized) {
-            initialize();
-        }
-        return pageAlignDirectMemory;
-    }
-
     private static void initialize() {
+        if (INIT_COUNT.get() == 2) {
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_internal_misc_VM.java]: With the updated patch it shouldn't be possible any more. Thanks!

**Comment by jerboaa** [comment]
Thanks for the review, @roberttoyonaga!

**Review by christianhaeubl — COMMENTED** [review]
I get the idea of the PR but I think it is too complex and at the same time not thread-safe enough. I implemented an alternative, see https://github.com/oracle/graal/pull/7553. Please let me know if that simpler approach is good enough.

**Comment by jerboaa** [comment]
> I get the idea of the PR but I think it is too complex and at the same time not thread-safe enough. I implemented an alternative, see #7553. Please let me know if that simpler approach is good enough.

Thanks. I'll take a look later today and let you know.

**Comment by jerboaa** [comment]
> Please let me know if that simpler approach is good enough.

It's good enough to unbreak us. Thanks! Closing as the replacement PR is #7553.

**Pull request closed without merging**